### PR TITLE
add_pool/k for k>2

### PR DESCRIPTION
**Pull request by jkingsbery**
First stab at some of the low hanging fruit on #111
- Refactor to use add_pool/2
- Refactored unit tests to primarily used add_pool/2
- Kept unit tests (and marked as deprecated in test name) for existing add_pool to ensure compatibility

A couple further suggestions:
- The environment test suite should be split into two: one suite for testing the creation of pools, and one for operations that use pools once created
- The default encoding for add_pool/2 was utf8, but based on the unit test the default encoding for the other add_pools was latin1. I had it as latin1 to keep existing unit tests happy, but it seems better for the default to be utf8.
- There could be an add_pool/1 that takes a pool record and creates it. I've refactored using an intermediate record to be able to use guards to check the types of things easily, but it would be better still if that was all done in add_pool/1 and add_pool/2 would just pass things in to the record.

### CODE REVIEW

**Comment by saa** [comment]
Why latin1 by default?

File: src/emysql.erl [hunk 1]
```diff
@@ -223,17 +223,57 @@ default_timeout() ->
 %%
 %% Refer to add_pool/8 for implementation details.
 
-add_pool(PoolId, Options) ->
-    Size = proplists:get_value(size, Options, 1),
+-record(pool_settings,{size,user,password,host,port,database,encoding,start_cmds,connect_timeout}).
```
Comment (jlouis) [inline src/emysql.erl:226]: We probably want this global in `emysql.hrl` so a name like `emysql_pool_settings` or `emysql_pool_options` or `emysql_pool_config` would be advantageous.

File: src/emysql.erl [hunk 2]
```diff
@@ -223,17 +223,57 @@ default_timeout() ->
...
+
+add_pool(PoolId, Options) when is_list(Options) ->
+    Size = proplists:get_value(size, Options, 5),
     User = proplists:get_value(user, Options, ""),
     Password = proplists:get_value(password, Options, ""),
     Host = proplists:get_value(host, Options, "127.0.0.1"),
     Port = proplists:get_value(port, Options, 3306),
     Database = proplists:get_value(database, Options, undefined),
-    Encoding = proplists:get_value(encoding, Options, utf8),
```
Comment (jlouis) [inline src/emysql.erl:233]: This change kinda irks me. On the other hand, this is MySQL and MySQL generally can't do utf8 right.

File: test/environment_SUITE.erl [hunk 3]
```diff
@@ -98,6 +110,8 @@ init_per_testcase(add_pool_env_all, Config) ->
     ok = application:start(emysql),
     Config;
 
+%TODO: Probably better to split out test suite into ones that expect the pool to exist and 
+% tests that are about setting up the pool.
```
Comment (jlouis) [inline test/environment_SUITE.erl:114]: Yes, definitely! And while we are at it, we should probably also merge the environment suite to the basics suite so we get proper coverage in there.

**Comment by jlouis** [comment]
This looks good. Some general comments:

The `latin1` vs `utf8` question is rather hard to decide. In a proper database system like Postgres, you just make the change and the database system does the-right-thing. In MySQL it is not so. First of all, utf8 support in MySQL uses at most 3 bytes. So 4-byte utf-8 encoded code point entities cannot be encoded in MySQL. They do exist. As soon as you go beyond the BMP (Basic Multilingual Plane) you have 4-byte codes. Emoji in Unicode 6.0 is a good example.

One has to be dead sure setting 'utf8' by default doesn't mess up stuff. On the other hand, this is what the Python connector driver does by default and I am inclined to say we should probably plan on doing the same.

I like the idea of `add_pool/1`. We should get that in. Out of this pull request springs two other requests:
- Add a `config_ok/1` call which can check the correctness of a server config and will `erlang:error(badarg)` if the configuration is not valid. Use this over the guards since it keeps correctness code in one place only.
- Since we are deprecating `add_pool/N` calls, don't use them in documentation anymore. Don't document the calls in `emysql` and keep them out of sight for newbies.

**Comment by jlouis** [comment]
Oh, and if you use utf8 in MySQL, your indexes are more limited in size than normally. There is also `utf8mb4` which can use 4 bytes, but then your index limitation is even smaller.

Mediawiki opted to store all utf8 blobs in binary for a reason :)

**Comment by jlouis** [comment]
Hmmm, we will need to keep this as latin1 for now. The other choice would break stuff all over the place in various ways. Yet another thing for a major update... _sigh_

**Pull request merged**